src/neo/Cryptography/MPT/MPTTrie.Proof.cs

```diff
@@ -0,0 +1,72 @@
+using Neo.IO;
+using Neo.Persistence;
+using System;
+using System.Collections.Generic;
+
+namespace Neo.Cryptography.MPT
+{
+    partial class MPTTrie<TKey, TValue>
+    {
+        public HashSet<byte[]> GetProof(TKey key)
+        {
+            var path = ToNibbles(key.ToArray());
+            if (path.Length < 1) return null;
+            HashSet<byte[]> set = new HashSet<byte[]>(ByteArrayEqualityComparer.Default);
+            if (!GetProof(ref root, path, set)) return null;
+            return set;
+        }
+
+        private bool GetProof(ref MPTNode node, ReadOnlySpan<byte> path, HashSet<byte[]> set)
+        {
+            switch (node)
+            {
+                case LeafNode leafNode:
+                    {
+                        if (path.Length < 1)
+                        {
+                            set.Add(leafNode.Encode());
+                            return true;
+                        }
+                        break;
+                    }
+                case HashNode hashNode:
+                    {
+                        if (hashNode.IsEmpty) break;
+                        var newNode = Resolve(hashNode);
+                        if (newNode is null) break;
+                        node = newNode;
+                        return GetProof(ref node, path, set);
+                    }
+                case BranchNode branchNode:
+                    {
+                        set.Add(branchNode.Encode());
+                        if (path.Length < 1)
+                        {
+                            return GetProof(ref branchNode.Children[BranchNode.ChildCount - 1], path, set);
+                        }
+                        return GetProof(ref branchNode.Children[path[0]], path[1..], set);
+                    }
+                case ExtensionNode extensionNode:
+                    {
+                        if (path.StartsWith(extensionNode.Key))
+                        {
+                            set.Add(extensionNode.Encode());
+                            return GetProof(ref extensionNode.Next, path[extensionNode.Key.Length..], set);
+                        }
+                        break;
+                    }
+            }
+            return false;
+        }
+
+        public static TValue VerifyProof(UInt256 root, TKey key, HashSet<byte[]> proof)
+        {
```

Comment:
It's possible to do this without the store?

Reply:
It uses a new `MemoryStore` in the next line.

Reply:
But this require to have an optimized `MemoryStore` and i remember a conversation when we decided to improve later `MemoryStore` because it was focused for developers, not for production, with this change it will be used in production.

@KickSeason please, don't resolve the conversation.

Reply:
Do you plan to switch to a real store in another PR?